### PR TITLE
38 product export

### PR DESCRIPTION
**Pull request by markuslamm**
please review @lauraluiz @schleichardt

### CODE REVIEW

File: app/inject/ProductProvider.java [hunk 1]
```diff
@@ -21,7 +22,7 @@
     private static final long ALLOWED_TIMEOUT = 3000;
 
     @Inject
-    public ProductProvider(final PlayJavaSphereClient playJavaSphereClient) {
+    public ProductProvider(final PlayJavaSphereClient playJavaSphereClient, final ImportService importService) {
```
Comment (schleichardt) [inline app/inject/ProductProvider.java]: Can you add here a comment why you have a dependency to `ImportService` but don't use it?

File: app/models/export/ProductDraftWrapper.java [hunk 2]
```diff
@@ -0,0 +1,266 @@
+package models.export;
```
Comment (schleichardt) [inline app/models/export/ProductDraftWrapper.java]: Important, in our company putting things into the sphere database is *import*. Rename this everywhere.

File: app/models/export/ProductDraftWrapper.java [hunk 3]
```diff
@@ -0,0 +1,266 @@
...
+
+    public ProductVariantDraftWrapper getMasterVariant() {
+        return masterVariant;
+    }
+
+    public List<ProductVariantDraftWrapper> getVariants() {
+        return variants;
+    }
+
+    public static class ProductVariantDraftWrapper {
```
Comment (schleichardt) [inline app/models/export/ProductDraftWrapper.java]: I would prefer this as normal class for SRP. https://en.wikipedia.org/wiki/Single_responsibility_principle

Comment (schleichardt) [inline app/models/export/ProductDraftWrapper.java]: This applies also for other nested classes.

File: conf/logback.xml [hunk 4]
```diff
@@ -9,7 +9,7 @@
     </appender>
 
     <appender name="FILE" class="ch.qos.logback.core.FileAppender">
-        <file>${application.home}/logs/donut.log</file>
```
Comment (schleichardt) [inline conf/logback.xml]: Why did you change this?

File: app/models/export/ProductDraftWrapper.java [hunk 5]
```diff
@@ -0,0 +1,266 @@
...
+        this.name = requireNonNull(name);
+        this.description = requireNonNull(description);
+        this.slug = requireNonNull(slug);
+        this.masterVariant = requireNonNull(masterVariant);
+        this.variants = requireNonNull(variants);
+        this.variants.add(masterVariant);
+    }
+
+    public ProductDraft createProductDraft(final ProductType productType, final TaxCategory taxCategory) {
+        final Reference<ProductType> productTypeReference = productType.toReference();
```
Comment (schleichardt) [inline app/models/export/ProductDraftWrapper.java]: I suggest to use directly `Reference<ProductType>` and `Reference<TaxCategory>` as parameters.

File: test/models/export/ProductDraftWrapperTest.java [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+
+    private ProductType productType;
+    private TaxCategory taxCategory;
+
+    @Before
+    public void setUp() {
+        productType = mock(ProductType.class);
+        final String productTypeId = UUID.randomUUID().toString();
+        when(productType.toReference()).thenReturn(Reference.of("product-type", productTypeId, productType));
+        when(productType.getId()).thenReturn(productTypeId);
```
Comment (schleichardt) [inline test/models/export/ProductDraftWrapperTest.java]: When you need to use mocks for your own code then this is a sign that sth. is not right with your API.

see https://github.com/commercetools/sphere-donut/pull/40/files#r41992292

File: test/models/export/ProductTypeDraftWrapperTest.java [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+    public void testCreateProductTypeDraft() throws Exception {
+        final ProductTypeDraftWrapper productTypeDraftWrapper =
+                JsonUtils.readObjectFromResource(PRODUCT_TYPE_JSON_RESOURCE, ProductTypeDraftWrapper.class);
+        assertThat(productTypeDraftWrapper).isNotNull();
+
+        final ProductTypeDraft productTypeDraft = productTypeDraftWrapper.createProductTypeDraft();
+        assertThat(productTypeDraft).isNotNull();
+        assertThat(productTypeDraft.getName()).isEqualTo("Donuts box");
+        assertThat(productTypeDraft.getDescription()).isEqualTo("A box with delicious donuts");
+        assertThat(productTypeDraft.getAttributes().size()).isEqualTo(5);
```
Comment (schleichardt) [inline test/models/export/ProductTypeDraftWrapperTest.java]: Can you also check the attributes?

File: test/services/ImportServiceTest.java [hunk 8]
```diff
@@ -0,0 +1,22 @@
+package services;
+
+
+import io.sphere.sdk.client.PlayJavaSphereClient;
+import org.junit.Before;
+import play.Configuration;
+
+import static org.mockito.Mockito.mock;
+
+public class ImportServiceTest {
```
Comment (schleichardt) [inline test/services/ImportServiceTest.java]: It has not test methods, so setUp is not called. Also assertions are missing.

File: conf/logback.xml (hunk 4 above)
Comment (markuslamm) [inline conf/logback.xml]: Sometimes, not every time,  ${application.home} which I thought was a default Logback variable, seemed not to be set correctly, and produced an additional "${application.home} _not_set" directory with the log files in my local environment.

File: README.md [hunk 9]
```diff
@@ -17,6 +17,13 @@ Visit a live demo of SPHERE-DONUT store at [iwantdonuts.com](http://www.iwantdon
 - Run `sbt run` command in root project directory.
 - Open your browser and point it to [http://localhost:9000](http://localhost:9000).
 
+### Product import
+
+There is a flag named `fixtures.import.enabled` in the configuration file `conf/application.conf`, which allows an initial import of the of the required, JSON-based product model
```
Comment (lauraluiz) [inline README.md]: `import of the of the required` typo :)

Comment (lauraluiz) [inline README.md]: Oh and unnecessary comma after that

File: README.md [hunk 10]
```diff
@@ -17,6 +17,13 @@ Visit a live demo of SPHERE-DONUT store at [iwantdonuts.com](http://www.iwantdon
...
+into the commercetools-platform. After the initial import succeeded, this flag can be set to `false` to improve the startup time of the application.
```
Comment (lauraluiz) [inline README.md]: Better `After a successful initial import, ...`

File: app/services/ImportServiceImpl.java [hunk 11]
```diff
@@ -0,0 +1,142 @@
...
+        if(importEnabled) {
+            importData();
+            importCustomType();
+        }
+    }
+
+    private void importData() {
+        LOG.debug("Starting Product import");
+        productExists().flatMap(productExits -> {
+            LOG.debug("Existing Product found: {}", productExits);
```
Comment (lauraluiz) [inline app/services/ImportServiceImpl.java]: `productExists`, I think ;)

File: app/services/ImportServiceImpl.java [hunk 12]
```diff
@@ -0,0 +1,142 @@
...
+
+    private F.Promise<Boolean> customTypeExists() {
+        final TypeQuery query = TypeQuery.of();
+        final F.Promise<PagedQueryResult<Type>> execute = playJavaSphereClient().execute(query);
+        return execute.map(typePagedQueryResult -> {
+            final Optional<Type> optionalType = typePagedQueryResult.getResults().stream()
+                    .filter(type -> CUSTOM_TYPE_KEY.equals(type.getKey())).findFirst();
+            return optionalType.isPresent();
+        });
+    }
```
Comment (lauraluiz) [inline app/services/ImportServiceImpl.java]: You can rewrite these methods in a simpler way:
```
private void importCustomType() {
    LOG.debug("Starting custom Type import");
    currentCustomType().onRedeem(currentCustomType -> {
        LOG.debug("Existing custom Type found: {}", currentCustomType.isPresent());
        if (!currentCustomType.isPresent()) {
            exportCustomType().onRedeem(type -> LOG.debug("Finished custom Type import, created '{}'", type.getKey()));
        }
    });
}

private F.Promise<Optional<Type>> currentCustomType() {
    final TypeQuery query = TypeQuery.of();
    final F.Promise<PagedQueryResult<Type>> execute = playJavaSphereClient().execute(query);
    return execute.map(typePagedQueryResult ->
            typePagedQueryResult.getResults().stream()
                .filter(type -> CUSTOM_TYPE_KEY.equals(type.getKey()))
                .findFirst()
    );
}
```

Comment (lauraluiz) [inline app/services/ImportServiceImpl.java]: I would rather return the promise of the custom type instead of losing it. This way it's up to the method calling the import to decide what to do with it. Also it makes the code easier to understand IMO.

```
    private F.Promise<Type> importCustomType() {
        LOG.debug("Starting custom Type import");
        return currentCustomType().flatMap(currentCustomType -> {
            LOG.debug("Existing custom Type found: {}", currentCustomType.isPresent());
            return currentCustomType
                    .map(type -> F.Promise.pure(currentCustomType.get()))
                    .orElseGet(() -> {
                        final F.Promise<Type> typePromise = exportCustomType();
                        typePromise.onRedeem(type -> LOG.debug("Finished custom Type import, created '{}'", type.getKey()));
                        return typePromise;
                    });
        });
    }

    private F.Promise<Optional<Type>> currentCustomType() {
        final TypeQuery query = TypeQuery.of();
        final F.Promise<PagedQueryResult<Type>> execute = playJavaSphereClient().execute(query);
        return execute.map(typePagedQueryResult ->
                typePagedQueryResult.getResults().stream()
                    .filter(type -> CUSTOM_TYPE_KEY.equals(type.getKey()))
                    .findFirst()
        );
    }
```

Comment (lauraluiz) [inline app/services/ImportServiceImpl.java]: The same for `importData`.

File: conf/data/product-draft.json [hunk 13]
```diff
@@ -0,0 +1,240 @@
...
+        "name": "quantity",
+        "value": 6
+      },
+      {
+        "name": "box",
+        "value": "Small box"
+      },
+      {
+        "name": "pactas1",
+        "value": "5231dfec1d8dd00f389dc547"
```
Comment (lauraluiz) [inline conf/data/product-draft.json]: All these Pactas IDs will only work with our Pactas account, which of course nobody else will have. We should let the developer provide these instead of giving them via the import. Maybe we can provide them via configuration file?
In any case, create better a ticket to improve import and let's solve this in another PR.

File: conf/logback.xml (hunk 4 above)
Comment (lauraluiz) [inline conf/logback.xml]: Not sure if we should push this change to the main branch instead of trying to find the root of the problem. You can always change your own version of logback.xml until then.

File: test/models/wrapper/ProductTypeDraftWrapperTest.java [hunk 14]
```diff
@@ -0,0 +1,85 @@
...
+        assertThat(productTypeDraft).isNotNull();
+        assertThat(productTypeDraft.getName()).isEqualTo("Donuts box");
+        assertThat(productTypeDraft.getDescription()).isEqualTo("A box with delicious donuts");
+        assertThat(productTypeDraft.getAttributes().size()).isEqualTo(5);
+    }
+
+    @Test
+    public void testQuantityAttributeDefinitions() {
+        checkAttributeAssertions("quantity", "Quantity", NumberType.of(), AttributeConstraint.COMBINATION_UNIQUE,
+                TextInputHint.SINGLE_LINE, Boolean.TRUE, Boolean.TRUE);
```
Comment (lauraluiz) [inline test/models/wrapper/ProductTypeDraftWrapperTest.java]: Why do you use the `Boolean` object instead of the primitive type?

File: it/services/ImportServiceIntegrationTest.java [hunk 15]
```diff
@@ -0,0 +1,76 @@
...
+import play.inject.guice.GuiceApplicationBuilder;
+
+import java.util.List;
+import java.util.Locale;
+import java.util.Optional;
+
+import static org.assertj.core.api.StrictAssertions.assertThat;
+import static play.test.Helpers.running;
+
+public class ImportServiceIntegrationTest {
```
Comment (lauraluiz) [inline it/services/ImportServiceIntegrationTest.java]: What's the blocker for these tests to work?
